### PR TITLE
KAFKA-7169: Validate SASL extensions through callback on server side

### PR DESCRIPTION
**Pull request by stanislavkozlovski**

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java
Comment (rondagostino) [inline clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java]: Collections.unmodifiableMap()

File: clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java
Comment (rondagostino) [inline clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java]: Objects.requireNonNull()

File: clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java
Comment (rondagostino) [inline clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java]: Objects.requireNonNull()

File: clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java
Comment (rondagostino) [inline clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java]: Maybe `inputExtensions()` is clearer?

File: clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java
Comment (rondagostino) [inline clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java]: extension name must not be empty

File: clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java
Comment (rondagostino) [inline clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java]: Maybe `valid(extensionName)` is better as it is the opposite of `error()` and not so close to `validatedExtensions()`?

File: clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java
Comment (rondagostino) [inline clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java]: Can initialize to `new HashMap<>()` here as is done with `invalidExtensions` below.

File: clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java
Comment (rondagostino) [inline clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java]: Can remove if initialize above

File: clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java
Comment (rondagostino) [inline clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java]: There should be a comment that as per the spec, unrecognized extensions should be ignored and neither of the above methods should be invoked for them.  We could even include `ignoredExtensions` -- @rajinisivaram?

File: clients/src/main/java/org/apache/kafka/common/security/oauthbearer/internals/OAuthBearerSaslServer.java
Comment (rondagostino) [inline clients/src/main/java/org/apache/kafka/common/security/oauthbearer/internals/OAuthBearerSaslServer.java]: %n adds a newline?  Not sure this is a good idea in a log message, best to make it a space, I think.

File: clients/src/test/java/org/apache/kafka/common/security/oauthbearer/internals/OAuthBearerSaslServerTest.java
Comment (rondagostino) [inline clients/src/test/java/org/apache/kafka/common/security/oauthbearer/internals/OAuthBearerSaslServerTest.java]: I'm wondering if `OAuthBearerUnsecuredValidatorCallbackHandler` should recognize all extensions as being valid.  This class is not meant for production use, and any extensions that are added in the JAAS config are meant to be there because they were explicitly put there for testing/development purposes, so it seems to make sense that it should just blindly mark everything as being valid.  Thoughts?

**Review by rondagostino — COMMENTED** [review]
Looks pretty good...

File: clients/src/test/java/org/apache/kafka/common/security/oauthbearer/internals/OAuthBearerSaslServerTest.java
Comment (stanislavkozlovski) [reply clients/src/test/java/org/apache/kafka/common/security/oauthbearer/internals/OAuthBearerSaslServerTest.java]: Yes, it does make sense. Will change

File: clients/src/main/java/org/apache/kafka/common/security/oauthbearer/internals/OAuthBearerSaslServer.java
Comment (stanislavkozlovski) [reply clients/src/main/java/org/apache/kafka/common/security/oauthbearer/internals/OAuthBearerSaslServer.java]: Why not, though?
A space could be better but my reasoning was that a new line makes it more clear

File: clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java
Comment (stanislavkozlovski) [reply clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java]: yeah

File: clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java
Comment (stanislavkozlovski) [reply clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java]: embarrassing :D

File: clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java
Comment (stanislavkozlovski) [reply clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java]: How about just `validate`?

File: clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java]: I think the name `validate` is confusing because it suggests that the actual validation is done by this method, while actually this method is simply marking the extension name valid. So perhaps just `valid` or something along those lines would be better. Javadoc also needs updating once the name is changed.

File: clients/src/main/java/org/apache/kafka/common/security/oauthbearer/internals/OAuthBearerSaslServer.java
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/common/security/oauthbearer/internals/OAuthBearerSaslServer.java]: Call we call this something other than `internal` error since that doesn't really convey much. Maybe even `handleCallbackError` would be better.

File: clients/src/main/java/org/apache/kafka/common/security/oauthbearer/internals/OAuthBearerSaslServer.java
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/common/security/oauthbearer/internals/OAuthBearerSaslServer.java]: Perhaps it was this way before, but we don't usually add `if (log.isDebugEnabled())` since it adds clutter without really providing sufficient advantage. Remove?

File: clients/src/main/java/org/apache/kafka/common/security/oauthbearer/internals/OAuthBearerSaslServer.java
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/common/security/oauthbearer/internals/OAuthBearerSaslServer.java]: Same as before, `if (log.isDebugEnabled())` not required.

File: clients/src/test/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallbackTest.java
Comment (rajinisivaram) [inline clients/src/test/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallbackTest.java]: We also need a test to validate that some extensions can be ignored (neither valid nor error).

File: clients/src/test/java/org/apache/kafka/common/security/oauthbearer/internals/OAuthBearerSaslServerTest.java
Comment (rajinisivaram) [inline clients/src/test/java/org/apache/kafka/common/security/oauthbearer/internals/OAuthBearerSaslServerTest.java]: Could we have this at top level and shared between test classes?

File: clients/src/test/java/org/apache/kafka/common/security/oauthbearer/internals/OAuthBearerSaslServerTest.java
Comment (rajinisivaram) [inline clients/src/test/java/org/apache/kafka/common/security/oauthbearer/internals/OAuthBearerSaslServerTest.java]: `it` => `SaslServer`

**Review by rajinisivaram — COMMENTED** [review]
@stanislavkozlovski Thanks for the PR, looks good, left a few minor comments.

File: clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java]: Yes, we could add `ignoredExtensions` and include that in the log in the server.

**Review by rajinisivaram — APPROVED** [review]
@stanislavkozlovski Thanks for the updates, LGTM. Will merge after @rondagostino gets a chance to review the updates.

File: clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java
Comment (stanislavkozlovski) [reply clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java]: If it's just for logging - we can make the server subtract `validatedExtensions()` from `inputExtensions()`. I don't feel it's needed

Comment (rondagostino) [reply clients/src/main/java/org/apache/kafka/common/security/oauthbearer/OAuthBearerExtensionsValidatorCallback.java]: @rajinisivaram @stanislavkozlovski LGTM with the possible exception of maybe adding support for retrieving/logging any ignored extensions?  I'll defer to your preference on this.

**Comment by rondagostino** [comment]
LGTM

**Comment by rajinisivaram** [comment]
@stanislavkozlovski Thanks for the updates, LGTM. Merging to trunk.

**Pull request merged**
